pkg/safeguards/preprocessing_helpers.go

```diff
@@ -0,0 +1,113 @@
+package safeguards
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"gopkg.in/yaml.v2"
+	"helm.sh/helm/v3/pkg/chart"
+	"helm.sh/helm/v3/pkg/chart/loader"
+	"helm.sh/helm/v3/pkg/chartutil"
+	"helm.sh/helm/v3/pkg/engine"
+)
+
+// Given a Helm chart directory or file, renders all templates and writes them to the specified directory
+func RenderHelmChart(isFile bool, mainChartPath, tempDir string) error {
+	if isFile { // Get the directory that the Chart.yaml lives in
+		mainChartPath = filepath.Dir(mainChartPath)
+	}
+
+	mainChart, err := loader.Load(mainChartPath)
+	if err != nil {
+		return fmt.Errorf("failed to load main chart: %s", err)
+	}
+
+	loadedCharts := make(map[string]*chart.Chart) // map of chart path to chart object
+	loadedCharts[mainChartPath] = mainChart
+
+	// Load subcharts and dependencies
+	for _, dep := range mainChart.Metadata.Dependencies {
+		// Resolve the chart path based on the main chart's directory
+		chartPath := filepath.Join(mainChartPath, dep.Repository[len("file://"):])
+		chartPath = filepath.Clean(chartPath)
+
+		subChart, err := loader.Load(chartPath)
+		if err != nil {
+			return fmt.Errorf("failed to load chart: %s", err)
+		}
+		loadedCharts[chartPath] = subChart
+	}
+
+	for chartPath, chart := range loadedCharts {
+		valuesPath := filepath.Join(chartPath, "values.yaml") // Enforce that values.yaml must be at same level as Chart.yaml
+		mergedValues, err := getValues(chart, valuesPath)
+		if err != nil {
+			return fmt.Errorf("failed to load values: %s", err)
+		}
+		e := engine.Engine{Strict: true}
+		renderedFiles, err := e.Render(chart, mergedValues)
+		if err != nil {
+			return fmt.Errorf("failed to render chart: %s", err)
+		}
+
+		if err := os.MkdirAll(tempDir, 0755); err != nil {
+			return fmt.Errorf("failed to create output directory: %s", err)
+		}
+
+		// Write each rendered file to the output directory with the same name as in templates/
+		for filePath, content := range renderedFiles {
+			outputFilePath := filepath.Join(tempDir, filepath.Base(filePath))
```

Comment:
This reads to me like it'll preserve the file structure that the helm charts existed in. We should think about how that will work with validation.

The validation code expects everything to be at the top level. i.e. `tempDir/deployment.yaml`, not `tempDir/subDir/deployment.yaml`.

Either we should make the validation code aware of this, or we should render the files directly under `tempDir`

Reply:
Yes-- we need all files to be written directly under the tempDir so that the validate command can process the manifests easily. All files are written directly under the tempDir, not a sub directory under the tempDir since it uses the base of the filepath, which is just the file name and no other directory. This is also verified in testing becuase we read and compare all rendered manifests from tempDir directly

Reply:
which portion of the validation code expects top-level manifest files? the manifest file loading is[ done recursively](https://github.com/Azure/draft/blob/fea83a411c111b4b343ebebe77452bf42848e242/pkg/safeguards/helpers.go#L211) at least

Reply:
The idea is for the render function to write all rendered manifests to the temporary directory at the same level (directly under it) regardless of where the template was found in the helm chart (main chart/ sub chart) so that the GetManifestFile() function can process it as it has been with a directory of .yamls.

Reply:
If it's done recursively, it should work either way. But this just makes it cleaner to test.